File: mobile/lib/features/character/domain/character_state.dart
/// F-042 캐릭터 현재 상태 (백엔드 GET /character/{idol_id}/state 응답).
///
/// row 없을 때 백엔드가 default(idle, 100/100/100)으로 채워줌.
//
// freezed 3 + redirected constructor에 @JsonKey 붙이면 invalid_annotation_target
// 경고가 뜨지만 generator는 정상 처리. 알려진 freezed 동작이라 파일 단위로 무시.
// ignore_for_file: invalid_annotation_target
library;

import 'package:freezed_annotation/freezed_annotation.dart';

import 'character_action.dart';
import 'furniture_placement.dart';

part 'character_state.freezed.dart';
part 'character_state.g.dart';

@freezed
abstract class CharacterState with _$CharacterState {
  const factory CharacterState({
    @JsonKey(name: 'idol_id') required String idolId,
    @JsonKey(name: 'current_action') required CharacterActionType currentAction,
    required int hunger,
    required int happiness,
    required int energy,
    @JsonKey(name: 'updated_at') required DateTime updatedAt,
    // PR-G2 — 드래그 위치(flame world 좌표). null = 미설정 → 기본 위치 사용.
    @JsonKey(name: 'position_x') double? positionX,
    @JsonKey(name: 'position_y') double? positionY,
    // 가구 배치 {kind: {x,y,w}}. null = 기본 배치(모바일 코드).
    @JsonKey(name: 'furniture_layout') Map<String, FurniturePlacement>? furnitureLayout,
  }) = _CharacterState;

  factory CharacterState.fromJson(Map<String, dynamic> json) =>
      _$CharacterStateFromJson(json);
}

File: backend/app/shared/models/character_state.py
# OWNER: character (F-042 캐릭터 상태)
"""character_states — 아이돌별 캐릭터 현재 상태 (1:1).

(idol_id) unique. 시간 경과/팬 행동/아이돌 트리거에 따라 변화.
"""

from datetime import datetime
from uuid import UUID

from sqlalchemy import DateTime, Float, ForeignKey, SmallInteger, text
from sqlalchemy import Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.types import Uuid

from app.core.db import Base
from app.shared.enums import CharacterActionType


class CharacterState(Base):
    __tablename__ = "character_states"

    id: Mapped[UUID] = mapped_column(
        Uuid, primary_key=True, server_default=text("gen_random_uuid()")
    )
    idol_id: Mapped[UUID] = mapped_column(
        Uuid,
        ForeignKey("profiles.id", ondelete="CASCADE"),
        nullable=False,
        unique=True,
    )
    current_action: Mapped[CharacterActionType] = mapped_column(
        SAEnum(
            CharacterActionType,
            name="character_action_type",
            create_type=False,
            values_callable=lambda e: [m.value for m in e],
        ),
        nullable=False,
        server_default="idle",
    )
    # 상태 수치 0-100. lazy 계산 우세 — read 시 시간 경과 반영.
    hunger: Mapped[int] = mapped_column(SmallInteger, nullable=False, server_default="100")
    happiness: Mapped[int] = mapped_column(SmallInteger, nullable=False, server_default="100")
    energy: Mapped[int] = mapped_column(SmallInteger, nullable=False, server_default="100")
    # PR-G2 드래그 위치 영속화 — flame world 좌표(logical). null = 미설정(모바일이 기본 위치 사용).
    position_x: Mapped[float | None] = mapped_column(Float, nullable=True)
    position_y: Mapped[float | None] = mapped_column(Float, nullable=True)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=text("NOW()"),
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=text("NOW()"),
    )
